javalib/src/main/scala/java/lang/Throwables.scala

```diff
@@ -49,49 +49,126 @@ private[lang] object StackTrace {
+
   def this() = this(null, null)
+
   def this(s: String) = this(s, null)
+
   def this(e: Throwable) = this(if (e == null) null else e.toString, e)
 
-  private var stackTrace: Array[StackTraceElement] = _
+  private[this] var stackTrace: Array[StackTraceElement] = _
 
-  fillInStackTrace()
+  if (writableStackTrace)
+    fillInStackTrace()
 
-  def initCause(cause: Throwable): Throwable = {
-    e = cause
-    this
-  }
+  // We use an Array rather than, say, a List, so that Throwable does not
+  // depend on the Scala collections.
+  private[this] var suppressed: Array[Throwable] = _
 
-  def getMessage(): String = s
+  final def addSuppressed(exception: Throwable): Unit = {
 
-  def getCause(): Throwable = e
+    if (exception == null) {
+      // Yes, JVM message uses 'cannot' (sic) & has terminal period/full_stop!
+      throw new java.lang.NullPointerException(
+        "Cannot suppress a null exception.")
+    }
 
-  def getLocalizedMessage(): String = getMessage()
+    if (exception == this) {
+      throw new java.lang.IllegalArgumentException(
+        "Self-suppression not permitted")
+    }
+
+    if (enableSuppression) this.synchronized {
+      if (suppressed eq null) {
+        suppressed = Array(exception)
+      } else {
+        val length        = suppressed.length
+        val newSuppressed = new Array[Throwable](length + 1)
+        System.arraycopy(suppressed, 0, newSuppressed, 0, length)
+        newSuppressed(length) = exception
+        suppressed = newSuppressed
+      }
+    }
+  }
 
   def fillInStackTrace(): Throwable = {
-    this.stackTrace = StackTrace.currentStackTrace()
+    // currentStackTrace should be handling exclusion in its own
+    // critical section, but does not. So do
+    if (writableStackTrace) this.synchronized {
+      this.stackTrace = StackTrace.currentStackTrace()
+    }
     this
   }
 
+  def getCause(): Throwable = e
+
+  def getLocalizedMessage(): String = getMessage()
+
+  def getMessage(): String = s
+
   def getStackTrace(): Array[StackTraceElement] = {
-    if (stackTrace eq null) {
-      stackTrace = Array.empty
-    }
-    stackTrace
+    // Be robust! Test this.stackTrace against null rather than relying upon
+    // the value of writableStackTrace.
+    //
+    // subclass scala.util.control.NoStackTrace overrides fillInStackTrace()
+    // so that it never touches this.stackTrace. This creates the situation
+    // where writeableStackTrace is true and this.stackTrace is null.
+    //
+    // If scala code creates this situation, then by the Gell-Mann principle
+    // user code in the wild is bound to do so.
+    //
+    // If stackTrace is null, no profit to calling fillInStackTrace().
+    // If writableStackTrace is true, then fillInStackTrace has already
+    // been called in the constructor. If the stack is not writable, then
+    // it can not be filled in.
+
+    if (stackTrace == null) {
+      Array.empty[StackTraceElement] // as specified by Java 8.
```

Comment:
Prefer
```suggestion
      new Array[StackTraceElement](0) // as specified by Java 8.
```
which does not depend on Scala's `ClassTag`s. We should avoid to depend on anything Scala-specific in `Throwable`, which is a core class.

Reply:
Can `new Array[StackTraceElement](0) ` be shared?

Reply:
I'm afraid not. On the JVM `getStackTrace()` always returns a new array, which is not `eq` to other 0-arg arrays (even coming from the same instance).

Reply:
Changed as recommended.  I will use that practice in javalib code from now on, including responding to review.